src/libmcrx.c

```diff
@@ -1055,3 +1075,387 @@ mcrx_is_system_error(enum mcrx_error_code err) {
   return err >= MCRX_ERR_SYSCALL_BIND;
 }
 
+/**
+ * mcrx_mnatmap_new:
+ * @mnatmapp: pointer to be filled with the new mnat context
+ *
+ * Create a new mnat context.
+ *
+ * Returns: error code
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnatmap_new(
+    struct mcrx_mnatmap **mnatmapp) {
+  if (mnatmapp == NULL) {
+    return MCRX_ERR_NULLARG;
+  }
+
+  struct mcrx_mnatmap *mnat_ctx;
+  mnat_ctx = calloc(1, sizeof(struct mcrx_mnatmap));
+  if (!mnat_ctx) {
+    return MCRX_ERR_NOMEM;
+  }
+
+  mnat_ctx->refcount = 1;
+  LIST_INIT(&mnat_ctx->mnats_head);
+
+  *mnatmapp = mnat_ctx;
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnatmap_apply:
+ * Apply the mnat context to mcrx context
+ *
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnatmap_apply(struct mcrx_ctx *ctx,
+    struct mcrx_mnatmap *mnatmap) {
+  if (ctx == NULL) {
+    return MCRX_ERR_NULLARG;
+  }
+  if (ctx->mnat_map == NULL) {
+    // no mnat running time mcrx context yet
+    mcrx_mnatmap_clone(mnatmap, &ctx->mnat_map);
```

Comment:
If this fails then we should fail with error code indicating out of memory.

If any of the new_entry allocations fail, we should also fail with out of memory, and we should do that before any unsubscriptions, in a sort of transactional model (so we don't have a half-applied mnatmap that aborts halfway through, or at least any cases that do that will report it as an internal error).